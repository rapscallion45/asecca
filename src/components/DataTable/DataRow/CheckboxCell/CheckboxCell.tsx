import { FC, useCallback, ChangeEvent, memo } from 'react';
import { Checkbox } from '@mui/material';
import { IDataTableEditCheckboxCellValueCallback } from '../../types';
import StyledTableCell from '../StyledCellWrapper/StyledCellWrapper';

/**
 * Data Table Checkbox Cell Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 *
 * @typedef ICheckboxCellProps
 * @prop {string} inputId - ID of the cell input field
 * @prop {boolean} canEdit - cell is editable flag
 * @prop {boolean | null} value - cell boolean value, can be null
 * @prop {IDataTableEditCheckboxCellValueCallback} submitCellValue - submit cell value

 * @prop {any} sx - cell styling overrrides
 */
interface ICheckboxCellProps {
  inputId: string;
  canEdit: boolean;
  value: boolean | null;
  submitCellValue?: IDataTableEditCheckboxCellValueCallback;
  sx?: any;
}

/**
 * Data Table Checkbox Cell
 *
 * Checkbox table cell component for displaying a table value as a boolean
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 *
 * @component
 * @param {ICheckboxCellProps} props - component props
 * @returns {FC} - data table checkbox cell functional component
 */
const Cell: FC<ICheckboxCellProps> = (props) => {
  const { inputId, canEdit, value, submitCellValue, sx } = props;

  /**
   * Callback for handling user input.
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.13
   *
   * @method
   * @param {ChangeEvent<HTMLInputElement>} event - value change trigger event
   */
  const handleValueChange = useCallback(
    (event: ChangeEvent<HTMLInputElement>) => {
      if (submitCellValue) submitCellValue(event.target.checked);
    },
    [submitCellValue]
  );

  return (
    <StyledTableCell align="left" sx={sx}>
      <Checkbox
        checked={value as boolean}
        onChange={handleValueChange}
        disabled={!canEdit}
        inputProps={{ 'aria-label': inputId }}
      />
    </StyledTableCell>
  );
};

export default memo(Cell);
