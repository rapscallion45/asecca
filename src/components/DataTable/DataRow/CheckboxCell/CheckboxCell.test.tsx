import React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Table, TableBody, TableRow } from '@mui/material';
import CheckboxCell from './CheckboxCell';

/* test values */
const testValue: boolean | null = true;

/* test style */
const testSx = { fontWeight: 'bold' };

/* mock callback for submit cell value */
const mockSubmitCellValueCallback = jest.fn((x: boolean | null) => {
  /* disable linter - do not care value is never read */
  /* eslint-disable */
  const mockStateValue: boolean | null = x;
  /* eslint-enable */
});

/**
 * Data Table Checkbox Cell Unit Tests
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.22
 */
describe('Data Table Checkbox Cell', () => {
  it('Renders correctly', async () => {
    /** perform snapshot test */
    const tree = renderer
      .create(
        <Table>
          <TableBody>
            <TableRow>
              <CheckboxCell
                canEdit
                inputId="testrow-testkey-input"
                value={testValue}
                sx={testSx}
              />
            </TableRow>
          </TableBody>
        </Table>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should submit value if cell clicked', async () => {
    /** Arrange */
    /** Act */
    render(
      <Table>
        <TableBody>
          <TableRow>
            <CheckboxCell
              canEdit
              inputId="testrow-testkey-input"
              value={testValue}
              sx={testSx}
              submitCellValue={mockSubmitCellValueCallback}
            />
          </TableRow>
        </TableBody>
      </Table>
    );

    /** Assert - checkbox should be rendered */
    expect(screen.queryByRole('checkbox')).toBeInTheDocument();

    /** Act - click on checkbox to toggle value */
    fireEvent(
      screen.getByRole('checkbox'),
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    );

    /** Assert - wait for input to toggle */
    await waitFor(() => {
      expect(screen.queryByRole('checkbox')).toHaveAttribute(
        'checked',
        testValue ? '' : 'checked'
      );
    });

    /** Assert - submit cell callback to have been called once, with toggled value */
    expect(mockSubmitCellValueCallback.mock.calls).toHaveLength(1);
    expect(mockSubmitCellValueCallback.mock.calls[0][0]).toBe(!testValue);
  });
});
