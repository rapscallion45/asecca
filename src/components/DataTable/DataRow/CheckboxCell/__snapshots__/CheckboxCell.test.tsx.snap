// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data Table Checkbox Cell Renders correctly 1`] = `
<table
  className="MuiTable-root css-rqglhn-MuiTable-root"
  role={null}
>
  <tbody
    className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
    role={null}
  >
    <tr
      className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
      role={null}
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1nt70hm-MuiTableCell-root"
      >
        <span
          className="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorSecondary Mui-checked MuiCheckbox-root MuiCheckbox-colorSecondary css-1sg8fd1-MuiButtonBase-root-MuiCheckbox-root"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={null}
        >
          <input
            aria-label="testrow-testkey-input"
            checked={true}
            className="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate={false}
            disabled={false}
            onChange={[Function]}
            required={false}
            type="checkbox"
          />
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
        </span>
      </td>
    </tr>
  </tbody>
</table>
`;
